packages/integration-objects/srv/integration-objects/simple-exchange-rate-type-service.js

```diff
@@ -0,0 +1,203 @@
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeUpdate(req) {
+    try {
+      await validatePayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function validatePayload(req) {
+    validateTenantIdInPayload(req);
+    validateFieldsForExchangeRateTypes(req.data);
+    await checkUniquenessForPrimaryKeys(req);
+  }
+
+  function beforeDelete(req) {
+    try {
+      validateDelete(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeCreateForRateTypeDescriptions(req) {
+    try {
+      await validateDescPayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeUpdateForRateTypeDescriptions(req) {
+    try {
+      await validateDescPayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function validateDescPayload(req) {
+    validateFieldsForExchangeRateTypeDescription(req.data, req.params);
+    await checkForValidityOfKey(req, ExchangeRateTypes);
+    await checkUniquenessForPrimaryKeysForExchangeRateTypeDescForUpdate(req);
+  }
+
+  async function checkUniquenessForPrimaryKeys(req) {
+    const affectedRows = await fetchExistingRateTypes(ExchangeRateTypes, req);
+    if (affectedRows.length) {
+      checkForDuplicateRateTypeObjects(affectedRows, req);
+    }
+  }
+
+  async function checkUniquenessForPrimaryKeysForExchangeRateTypeDescForUpdate(req) {
+    const affectedRows = await fetchExistingRateTypeDesc(ExchangeRateTypeDescriptions, req);
+    if (affectedRows.length) {
+      checkForDuplicateRateTypeObjects(affectedRows, req);
+    }
+  }
+
+  async function beforeActiveDeleteForExchangeRateTypeDescription(req) {
+    const record = req.data;
+    const affectedRows = await fetchExistingTexts(ExchangeRateTypeDescriptions, record, req);
+    if (affectedRows.length && !util.isNullish(affectedRows[0].ID_texts)) {
+      await validateIdInTenant(affectedRows[0].ID_texts, ExchangeRateTypes, req, affectedRows);
+    }
+  }
+
+  async function writeAuditLogForRateTypes(req, next) {
+    let currentValues;
+    const data = req.data;
+    const type = 'exchange-rate-type';
+    const auditParams = {
+      exchangeRateType: data.exchangeRateType,
+      exchangeRateTypeDescription: data.exchangeRateTypeDescription,
+      referenceCurrencyThreeLetterISOCode: data.referenceCurrencyThreeLetterISOCode
+    };
+    let currentValuesIsNotEmpty = true;
+
+    if (Constants.CREATE_EVENT !== req.event) {
+      currentValues = await fetchCurrentAttributeValues(req, req.user.tenant, ExchangeRateTypes);
+      currentValuesIsNotEmpty = !util.isNullish(currentValues);
+    }
+
+    if (currentValuesIsNotEmpty) {
+      await writeAuditLog(req, auditParams, currentValues, type);
+    }
+
+    await next();
+  }
```

Comment:
I understand that you want log in the following cases:
1. `req.event === Constants.CREATE_EVENT` and `currentValues === undefined`
2. `req.event !== Constants.CREATE_EVENT` and `currentValues !== undefined`

That also means that you don't write audit logs in this case:
`req.event !== Constants.CREATE_EVENT` and `currentValues === undefined`
I assume this is intended behavior. This is very difficult to read in the existing code, especially because `currentValuesIsNotEmpty` is `true` by default when `currentValues` is `undefined`, but will become `false` when after fetching `currentValues` is nullish.
Consider the following. I duplicated the call to `writeAuditlog`, but it is more explicit and therefore easier to understand and less errorprone. Feel free to improve further.
```suggestion
async function writeAuditLogForRateTypes(req, next) {
    const data = req.data;
    const type = 'exchange-rate-type';
    const auditParams = {
      exchangeRateType: data.exchangeRateType,
      exchangeRateTypeDescription: data.exchangeRateTypeDescription,
      referenceCurrencyThreeLetterISOCode: data.referenceCurrencyThreeLetterISOCode
    };
    
    if (req.event === Constants.CREATE_EVENT) {
      await writeAuditLog(req, auditParams, undefined, type);
    } else {
      const currentValues = await fetchCurrentAttributeValues(req, req.user.tenant, ExchangeRateTypes);
      if (!util.isNullish(currentValues)) {
        await writeAuditLog(req, auditParams, currentValues, type);
      }
    }
    await next();
  }
};
```

Reply:
This is done